src/main.cpp

```diff
@@ -336,6 +336,24 @@ void MarkBlockAsInFlight(NodeId nodeid, const uint256& hash, CBlockIndex *pindex
     mapBlocksInFlight[hash] = std::make_pair(nodeid, it);
 }
 
+// Requires cs_main.
+bool IsBlockTransferTimedOut(const uint256& hash, NodeId newnode) {
+    // 20 minutes ~ 2 blocks - maybe this should be configurable for slow nodes
+    int64_t nTimeout = GetTimeMicros() - Params().TargetSpacing() * 2 * 1000000;
+
+    map<uint256, pair<NodeId, list<QueuedBlock>::iterator> >::iterator itInFlight = mapBlocksInFlight.find(hash);
+    if (itInFlight->second.second->nTime < nTimeout &&
+        itInFlight->second.first != newnode)
+    {
+        LogPrintf("Peer=%d failed to transfer block %s in twice the expected block interval, requesting elsewhere...\n",
+                  itInFlight->second.first, hash.ToString());
+        Misbehaving(itInFlight->second.first, 10);
```

Comment:
I think we want to disconnect such a node (but without raising banning score highly) immediately, to avoid having them think we still want the block (as there is no way to cancel a download request).